File: LeetCode/JavaScript/#543 Diameter of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

var diameterOfBinaryTree = function(root) {
   if(root===null||root.left===null&&root.right===null)return 0;
   var res = 0;
   function deep(root){
        var num1 = 0;
        var num2 = 0;
        if(root===null) return 0;
        num1 = deep(root.left);
        num2 = deep(root.right);
        res = Math.max(res,num1+num2);
        return Math.max(num1,num2)+1;
   }
   deep(root);
   return res;
};

File: LeetCode/Python/#543 Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        self.ans = 0
        def deep(root):
            if not root:
                return 0
            l = deep(root.left)
            r = deep(root.right)
            self.ans = max(self.ans,l+r)
            return max(l,r)+1
        deep(root)
        return self.ans
